build-support/php: make validation check a non blocking process

drupol opened the pull request:
This PR:

1. Add the color red for error messages, orange for warnings. Exemple:
  ![image](https://github.com/NixOS/nixpkgs/assets/252042/185ead40-167f-4eab-8937-2da4b75900e2)
2. Make the `composer validate` a step that can be skipped by the user by adding the `composerStrictValidation` set to `false`; by default and to preserve the current behaviour, it is set to `true`. 
3. The Grumphp update has been removed - after thinking twice, I won't provide the update until the upstream repo is updated.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

drupol committed:
build-support/php: add `composerStrictValidation` attribute

Set to `true` by default

drupol committed:
build-support/php: add colors in case of errors

yu-re-ka commented on pkgs/build-support/php/hooks/composer-repository-hook.sh:
```diff
@@ -65,7 +65,16 @@ composerRepositoryBuildHook() {
     echo "Executing composerRepositoryCheckHook"
 
-    composer validate --no-ansi --no-interaction
+    if ! composer validate --strict --no-ansi --no-interaction --quiet; then
```
Are you _sure_ that on mismatch it will not silently update the composer.lock and download a random/latest version of the dependencies?

drupol replied:
Yes 100% sure. It won't alter the reproducible behaviour since Composer basically download what's in `composer.lock` only.

yu-re-ka reviewed (commented):
In my opinion it would be safer to rely on patches in these edge cases. That means for each application that ships an inconsistent/outdated composer.lock, add a patch that brings composer.json and composer.lock into a consistent state.

drupol commented:
I modified the PR and added a new attribute to the PHP/Composer builder: `composerStrictValidation` which is set to `true` by default to preserve the actual behaviour. It can now be overidden by the user.

In case of error, the process displays a message in red and stop the build.
In case of `composerStrictValidation = false;` and discrepancies between `composer.lock` and `composer.json`, it displays a warning in orange, and the build continue.

Ma27 commented:
IIRC the same happens with Rust: if the lockfile is out of date, the build breaks. This is IMHO an upstream bug then and should be fixed upstream and patched away until then.

Also, what's the point of a validation if it's optional?

drupol commented:
It's enabled by default, but can be disabled (and display a message to the user).

In some situations such discrepancies may happen, have a look at the issue I opened on the project `grumphp-shim`.

Fortunately, such things won't break the build and that's the reason why I made it optional, but enabled by default.
